### PR TITLE
port: Support object comparison in ==

### LINKED ISSUE #3671: port: Support object comparison in == (#5575)
The changes in [Support object comparison in == (#5575)](https://github.com/microsoft/botbuilder-dotnet/pull/5575) may need to be ported to maintain parity with `microsoft/botbuilder-dotnet`.

<blockquote>
Fixes #5563

## Description
Support object comparison in `==`

## Specific Changes
1. Support object comparison in `contains`
2. Support object comparison for operator `==` and `!=`
</blockquote>

Please review and, if necessary, port the changes.

### PR DESCRIPTION
**Pull request by Danieladu**
Fixes #3671

## Description
Support object comparison in `==`

## Specific Changes
1. Support object comparison in `contains`
2. Support object comparison for operator `==` and `!=`

### CODE REVIEW

**Review by joshgummersall — APPROVED** [review]

File: libraries/adaptive-expressions/src/builtinFunctions/contains.ts [hunk 1]
```diff
@@ -35,10 +35,15 @@ export class Contains extends ExpressionEvaluator {
         const { args, error: childrenError } = FunctionUtils.evaluateChildren(expression, state, options);
         let error = childrenError;
         if (!error) {
-            if ((typeof args[0] === 'string' && typeof args[1] === 'string') || Array.isArray(args[0])) {
+            if (typeof args[0] === 'string' && typeof args[1] === 'string') {
                 found = args[0].includes(args[1]);
-            } else if (args[0] instanceof Map) {
```
Comment (joshgummersall) [inline libraries/adaptive-expressions/src/builtinFunctions/contains.ts:40]: Does it seem like this logic is getting lost in this refactor?

File: libraries/adaptive-expressions/src/functionUtils.ts [hunk 2]
```diff
@@ -768,14 +770,43 @@ export class FunctionUtils {
...
-        if (Array.isArray(obj1) && obj1.length === 0 && Array.isArray(obj2) && obj2.length === 0) {
-            return true;
+        // Array Comparison
+        if (Array.isArray(obj1) && Array.isArray(obj2)) {
+            if (obj1.length !== obj2.length) {
+                return false;
+            }
+
+            let isEqual = true;
+            for (let i = 0; i < obj1.length; i++) {
```
Comment (joshgummersall) [inline libraries/adaptive-expressions/src/functionUtils.ts:780]: Could simplify this (incorporating early return) by writing
```typescript
return obj1.every((item, i) => FunctionUtils.commonEquals(item, obj2[i]));
```

File: libraries/adaptive-expressions/src/functionUtils.ts [hunk 3]
```diff
@@ -768,14 +770,43 @@ export class FunctionUtils {
...
+        // Object Comparison
+        const propertyCountOfObj1 = FunctionUtils.getPropertyCount(obj1);
+        const propertyCountOfObj2 = FunctionUtils.getPropertyCount(obj2);
+        if (propertyCountOfObj1 >= 0 && propertyCountOfObj2 >= 0) {
+            if (propertyCountOfObj1 !== propertyCountOfObj2) {
+                return false;
+            }
+            let jsonObj1 = obj1;
+            let jsonObj2 = obj2;
+            if (obj1 instanceof Map) {
```
Comment (joshgummersall) [inline libraries/adaptive-expressions/src/functionUtils.ts:798]: Maybe this should be pushed into the `convertToObj` method, so these lines become `const jsonObj1 = FunctionUtils.convertToObj(obj1);`

File: libraries/adaptive-expressions/src/functionUtils.ts [hunk 4]
```diff
@@ -826,4 +857,37 @@ export class FunctionUtils {
...
+                const obj = {};
+                for (const [key, value] of instance) {
+                    obj[key] = FunctionUtils.convertToObj(value);
+                }
+                return obj;
+            } else {
+                // Convert Object recursionly
+                const objInstance = instance as Record<string, unknown>;
+                const obj = {};
+                for (const key in objInstance) {
```
Comment (joshgummersall) [inline libraries/adaptive-expressions/src/functionUtils.ts:877]: Probably want `key of` here.

File: libraries/adaptive-expressions/src/functionUtils.ts [hunk 5]
```diff
@@ -826,4 +857,37 @@ export class FunctionUtils {
...
+                }
+                return obj;
+            }
+        } else if (Array.isArray(instance)) {
+            // Convert Array
+            const result = [];
+            for (const item of instance) {
+                result.push(FunctionUtils.convertToObj(item));
+            }
+            return result;
```
Comment (joshgummersall) [inline libraries/adaptive-expressions/src/functionUtils.ts:888]: `return instance.map(item => FunctionUtils.convertToObj(item))`

File: libraries/adaptive-expressions/src/builtinFunctions/contains.ts (hunk 1 above)
Comment (Danieladu) [reply libraries/adaptive-expressions/src/builtinFunctions/contains.ts:40]: Yes. The logic of `contains` of Map would be handlered in `accessProperty`

File: libraries/adaptive-expressions/src/functionUtils.ts (hunk 2 above)
Comment (Danieladu) [reply libraries/adaptive-expressions/src/functionUtils.ts:780]: Good suggestion!

File: libraries/adaptive-expressions/src/functionUtils.ts [hunk 6]
```diff
@@ -826,4 +843,33 @@ export class FunctionUtils {
 
         return count;
     }
+
+    /**
+     * @private
+     */
+    private static convertToObj(instance: unknown) {
+        if (FunctionUtils.getPropertyCount(instance) >= 0) {
+            if (instance instanceof Map) {
```
Comment (joshgummersall) [inline libraries/adaptive-expressions/src/functionUtils.ts:852]: Last suggestion, then I promise I won't code-golf anymore on this PR 😄 

```typescript
const entries = instance instanceof Map ? instance : Object.entries(instance);
return entries.reduce((acc, [key, value]) => ({...acc, [key]: FunctionUtils.convertToObj(value)}), {});
```

Comment (Danieladu) [reply libraries/adaptive-expressions/src/functionUtils.ts]: More suggestions better! Means more can learn.
`reduce` could not apply on `IterableIterator` directly. So I use Array.from(instance.entries()) to convert it into Array.

**Review by joshgummersall — APPROVED** [review]

**Comment by Danieladu** [comment]
The browserBot is always failing. But I could not access the details of the error message. @joshgummersall  Could you help to take a look and whether some APIs are unavailable.

**Comment by joshgummersall** [comment]
@Danieladu yep I discovered that myself today! I'll land this and try to resolve it tomorrow.

**Pull request merged**